viewing_party/party.py

```diff
@@ -1,23 +1,129 @@
+            user_data["watchlist"].remove(movie)
+            break    
+    return user_data
+
+    
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    sum = 0.0
+    user_data_len = len(user_data["watched"])
+    if user_data_len >0:
+        for movie in user_data["watched"]:
+            sum += movie["rating"]
+
+        return sum/user_data_len
+    else:
+        return sum
+
 
+def get_most_watched_genre(user_data):
+    most_watched_dict = {}
+    user_data_len = len(user_data["watched"])
+
+    if user_data_len > 0:
+        for movie in user_data["watched"]:
+            if movie["genre"] in most_watched_dict:
+                most_watched_dict[movie["genre"]] += 1
+            else:
+                most_watched_dict[movie["genre"]] = 0
+
+        most_watched_genre = max(most_watched_dict, key=most_watched_dict.get)
+        return most_watched_genre
+    else:
+        return None
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    unique_movies = []
+    friends_movies = set()
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_movies.add(movie["title"])
 
+    for user_movie in user_data["watched"]:
+        if user_movie["title"] not in friends_movies:
+            unique_movies.append(user_movie)
+
+    return unique_movies
         
+def get_friends_unique_watched(user_data):
+    unique_movies = []
+    user_movies = []
+
+    for user_movie in user_data["watched"]:
+        user_movies.append(user_movie["title"])
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] not in user_movies and movie not in unique_movies:
+                unique_movies.append(movie)
+
+    return unique_movies
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    friends_recs = []
+    friends_unique_data = get_friends_unique_watched(user_data)
+
+    if len(friends_unique_data) >0 and len(user_data["watched"])> 0:
+        for movie in friends_unique_data:
+            if movie["host"] in user_data["subscriptions"]:
+                friends_recs.append(movie)
+
+    return friends_recs
+
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    rec_by_genre = []
+    most_watched_genre = get_most_watched_genre(user_data)
+    recs = get_available_recs(user_data)
+    for movie in recs:
+        if movie["genre"] == most_watched_genre:
+            rec_by_genre.append(movie)
+
+    return rec_by_genre

```

Comment:
Though this solution passes the current test suite, it isn't doing what the problem statement asks. 

The test `test_new_genre_rec` uses `clean_wave_5_data()` as the input, which includes subscription data. If we remove "hulu" from the user's subscriptions on line 157 of `test_constants.py`, suddenly the test `test_new_genre_rec` starts failing. Our problem statement for `get_new_rec_by_genre` does not ask us to consider subscriptions though, so changing that subscription data should not change the result of `get_new_rec_by_genre`. 

- What steps can you take to diagnose what is happening? 
- What was the cause of the issue?